lib/service/awsoidc.go

```diff
@@ -0,0 +1,414 @@
+		Issuer: updater.AWSOIDCProviderAddr,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	req := &awsoidc.AWSClientRequest{
+		IntegrationName: integration.GetName(),
+		Token:           token,
+		RoleARN:         integration.GetAWSOIDCIntegrationSpec().RoleARN,
+		Region:          awsRegion,
+	}
+
+	// The deploy service client is initialized using AWS OIDC integration.
+	deployServiceClient, err := awsoidc.NewDeployServiceClient(ctx, req, updater.AuthClient)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// ownershipTags are used to identify if the ecs resources are managed by the
+	// teleport integration.
+	ownershipTags := map[string]string{
+		types.ClusterLabel:     updater.TeleportClusterName,
+		types.OriginLabel:      types.OriginIntegrationAWSOIDC,
+		types.IntegrationLabel: integration.GetName(),
+	}
+
+	// Acquire a lease for the region + integration before attempting to update the deploy service agent.
+	// If the lease cannot be acquired, the update is already being handled by another instance.
+	semLock, err := updater.AuthClient.AcquireSemaphore(ctx, types.AcquireSemaphoreRequest{
+		SemaphoreKind: types.SemaphoreKindConnection,
+		SemaphoreName: fmt.Sprintf("update_deploy_service_agents_%s_%s", awsRegion, integration.GetName()),
+		MaxLeases:     1,
+		Expires:       updater.Clock.Now().Add(updateDeployAgentsInterval),
+		Holder:        "update_deploy_service_agents",
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), teleport.MaxLeases) {
+			updater.Log.WithError(err).Debug("Deploy service agent update is already being processed.")
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+	defer func() {
+		if err := updater.AuthClient.CancelSemaphoreLease(ctx, *semLock); err != nil {
+			updater.Log.WithError(err).Error("Failed to cancel semaphore lease.")
+		}
+	}()
+
+	updater.Log.Debugf("Updating Deploy Service Agents for integration %s in AWS region: %s", integration.GetName(), awsRegion)
+	if err := awsoidc.UpdateDeployServiceAgent(ctx, deployServiceClient, awsoidc.UpdateServiceRequest{
+		TeleportClusterName: updater.TeleportClusterName,
+		TeleportVersionTag:  teleportVersion,
+		OwnershipTags:       ownershipTags,
+	}); err != nil {
+
+		switch {
+		case trace.IsNotFound(err):
+			// The updater checks each integration/region combination, so
+			// there will be regions where there is no ECS cluster deployed
+			// for the integration.
+			updater.Log.WithError(err).Debugf("Integration %s does not manage any services within region %s.", integration.GetName(), awsRegion)
+			return nil
+		case trace.IsAccessDenied(awslib.ConvertIAMv2Error(trace.Unwrap(err))):
+			// The aws oidc role may lack permissions due to changes in teleport.
+			// In this situation users should be notified that they will need to
+			// re-run the deploy service iam configuration script and update the
+			// permissions.
+			updater.Log.WithError(err).Warning("Re-run deploy service configuration script to update permissions.")
```

Comment:
Can we make this error message more useful? If I saw it, I would have no idea what it means and what I need to do.

Reply:
I've simplified the message to "Update integration role and add missing permissions.". The error already includes context on relevant role and operation 
```
error ECS: DescribeTaskDefinition, https response error StatusCode: 400, RequestID: 57f3de8c-3f42-4081-a004-d789d41f6475, api error AccessDeniedException: User: arn:aws:sts::278576220453:assumed-role/bernard-dev-database-access/1694123674929138369 is not authorized to perform: ecs:DescribeTaskDefinition on resource: * because no identity-based policy allows the ecs:DescribeTaskDefinition action
```